plugins/last-fm.py

```diff
@@ -0,0 +1,237 @@
+# last.fm plugin
+#
+# usage:
+# .np [lastfm_nickname] --- check what is playing right now for an user
+# NOTE: this bot saves the last lastfm nick you used. You *really* only
+# need to use this command like that once, afterwards you can just do
+# .np , and the bot will remember the lastfm nick associated with your
+# irc nick.
+# .np @[irc_username] --- check what's playing for an another user (if
+# they have their nicks saved)
+#
+#
+# TODO:
+# .compat [lastfm_nickname] [lastfm_nickname] --- get compatiblity
+# between two lastfm users.
+# NOTE .compat can be used like this too:
+# .compat @[irc_nick] (compares your saved nick with the argument nick)
+# .compat [lastfm_nick] (same as above, but with a lastfm nick)
+# .compat @[irc_nick] [lastfm_nick]
+# .compat [irc_nick] @[lastfm_nick]
+# .compat @[irc_nick] @[irc_nick]
+
+
+from typing import List
+import asyncio
+import requests
+from datetime import datetime
+
+# import core db functions
+from core import db, hook
+
+# db table definition
+quote_columns: List[str] = [
+    'irc_nick', 'lastfm_nick']
+
+api_url = 'http://ws.audioscrobbler.com/2.0/?format=json'
+
+
+def _get_lastfm_nick(conn, irc_nick):
+    nicks = db.get_row(conn, 'lastfm', 'irc_nick', irc_nick)
+    try:
+        return nicks[0][1]
+    except IndexError:
+        pass
+
+
+def _add_lastfm_nick(conn, irc_nick, lastfm_nick):
+    nick_check = None
+    nick_check = _get_lastfm_nick(conn, irc_nick)
+
+    if nick_check is None:
+        nickdata = (irc_nick, lastfm_nick)
+        db.set_row(conn, 'lastfm', nickdata)
+    else:
+        db.set_cell(conn, 'lastfm',
+                    'lastfm_nick', lastfm_nick,
+                    'irc_nick', irc_nick)
+
+    return lastfm_nick
+
+
+def _pull_latest_track(api_key, lastfm_nick):
+    url = api_url+f'&method=user.getrecenttracks'
+    url += f'&user={lastfm_nick}&api_key={api_key}'
+    response = requests.get(url)
+    if response.status_code == 200:
+        try:
+            track = response.json()['recenttracks']['track'][0]
+            return track
+        except IndexError:
+            pass
+
+
+def _pull_track_tags(api_key, track_title, track_artist):
+    url = api_url+f'&method=track.gettoptags&track={track_title}'
+    url += f'&artist={track_artist}&api_key={api_key}'
+    response = requests.get(url)
+    if response.status_code == 200:
+        try:
+            tags_json = response.json()['toptags']['tag'][0:5]
+        except IndexError:
+            tags_json = response.json()['toptags']['tag']
+
+        if len(tags_json) < 1:
+            return 'none'
+
+        tags = []
```

Comment:
`tags = ' , '.join(tag['name'] for tag in tags_json)` should work, or `tags = ' , '.join([tag['name'] for tag in tags_json])` if join doesn't like the generator expression, I forget if it cares.